src/mon/Elector.cc

```diff
@@ -381,10 +418,101 @@ void Elector::handle_nak(MonOpRequestRef op)
   // the end!
 }
 
+void Elector::begin_peer_ping(int peer)
+{
+  if (live_pinging.count(peer)) {
+    return;
+  }
+
+  dout(5) << __func__ << " against " << peer << dendl;
+
+  peer_tracker.report_live_connection(peer, 0); // init this peer as existing
+  live_pinging.insert(peer);
+  peer_acked_ping[peer] = ceph_clock_now();
+  send_peer_ping(peer);
+  mon->timer.add_event_after(ping_timeout / PING_DIVISOR,
+			     new C_MonContext{mon, [this, peer](int) {
+				 ping_check(peer);
+			       }});
+}
+
+void Elector::send_peer_ping(int peer, const utime_t *n)
+{
+  dout(10) << __func__ << " to peer " << peer << dendl;
+
+  utime_t now;
+  if (n != NULL) {
+    now = *n;
+  } else {
+    now = ceph_clock_now();
+  }
+  MMonPing *ping = new MMonPing(MMonPing::PING, now, peer_tracker.get_encoded_bl());
+  mon->messenger->send_to_mon(ping, mon->monmap->get_addrs(peer));
+  peer_sent_ping[peer] = now;
+}
+
+void Elector::ping_check(int peer) {
+  dout(20) << __func__ << " to peer " << peer << dendl;
+  utime_t now = ceph_clock_now();
+  utime_t& acked_ping = peer_acked_ping[peer];
+  utime_t& newest_ping = peer_sent_ping[peer];
+  if (!acked_ping.is_zero() && acked_ping < now - ping_timeout) {
+    peer_tracker.report_dead_connection(peer, now - acked_ping);
+    acked_ping = now;
+    live_pinging.erase(peer);
+    return;
```

Comment:
> it seems odd to stop pinging as soon as the peer is dead - it seems like this would cause the score to only decrease a small amount a single time when the timeout is reached. It's treated as 0 for the purposes of get_total_connection_scores(), but if the rank comes back the score will increase again.

Oh dear, I busted this! Good catch. The intent was to mark the connection dead every 2 seconds with that timer event a bit lower. o_0

> The first ping that's acked after it comes back will have a disproportionate effect on the score as well, increasing it very quickly since the prior acked ping time could be very old.

That shouldn't happen; when we begin_peer_ping() we set the peer_acked_ping to ceph_clock_now() so it will only add up to 2 seconds of liveness to the score. (And if the peer sends a PING_REPLY to an outdated ping, we invoke begin_peer_ping() to reset things.)


Does the scoring make sense if I fix that dead-not-being-counted bug?

Reply:
> > The first ping that's acked after it comes back will have a disproportionate effect on the score as well, increasing it very quickly since the prior acked ping time could be very old.
> 
> That shouldn't happen; when we begin_peer_ping() we set the peer_acked_ping to ceph_clock_now() so it will only add up to 2 seconds of liveness to the score. (And if the peer sends a PING_REPLY to an outdated ping, we invoke begin_peer_ping() to reset things.)

I see, I misread that.

> Does the scoring make sense if I fix that dead-not-being-counted bug?

Yes, it makes sense to me now. Thanks for clearing that up!

Reply:
...and now done.